spec/test_doubles.rb

```diff
@@ -123,6 +123,31 @@ class MakesTeaWithMilkAction
     end
   end
 
+  class MakesTeaMaybeWithMilkAction
+    extend LightService::Action
+    expects :tea, :milk, :use_milk
+    promises :milk_tea
+    maybe :milk
```

Comment:
This observation is about the DSL:

My first intuition was that the `maybe` macro is at the same level as where `expects` and `promises` are. I realized they are not when I removed `:milk` from the expected keys. I have to expect a key that might be listed in `:maybe`.

It seems the `:maybe` is coupled to the `:expects` macro, I wonder if it could be expressed as such.
What do you think about this:
```ruby
class MakesTeaMaybeWithMilkAction
  extend LightService::Action
  expects :tea, {:maybe => :milk}, :use_milk
  promises :milk_tea
  ...
end
```

Or if you want the distinction, I should be able to list keys that only `:maybe` used and not fully expect it.

What do you think?

Reply:
Also, would `:maybe` work on `:promised` keys as well?

Reply:
I agree that having to list the keys in two different places isn't the best approach here.  I think that what makes it tricky is that this PR is introducing a new concept.  Originally, `expects` just meant "This key needs to be in the context when this action is executed."  But with this change, it now means "This key needs to be in the context and must be accessed when the action is executed" and `maybe` means "This key needs to be in the context and might be accessed when the action is executed."  So we need to find a way to cleanly combine those concepts.

There's also a question of, if a key is expected but not used, does it have to exist in the context?  I think that the answer is yes, we want to be as explicit as possible about the inputs and outputs of an action.  In that light, the term `maybe` might cause confusion, because it could imply that a key doesn't have to be present in the context.  I don't have a good alternative, though.  I've thought of `conditional`, `sometimes`, `ignore` and `maybe_used`.  (`Maybe` also has an entirely different meaning in Haskell, which lead to confusion as well.)

A couple of possibilities for improving the syntax (I'll keep using `maybe` here for consistency):
* Completely separate `expects` and `maybe`, each key is declared in one or the other, but not both.  We would verify that both types of keys exist at the start of execution but would only verify that `expects` keys were accessed during execution.  The above example would become:
```
class MakesTeaMaybeWithMilkAction
  extend LightService::Action
  expects :tea, :use_milk
  maybe :milk
  promises :milk_tea
  ...
end
```

* Let keys be declared as hashes, as you suggested.  I think it might be cleaner to swap your suggested key and value, though:
```
class MakesTeaMaybeWithMilkAction
  extend LightService::Action
  expects :tea, {:milk => :maybe}, :use_milk
  promises :milk_tea
end
```

* Add `:maybe` as a keyword arg to the end of the `expects` declaration.  I think that I like this syntax the best:
```
class MakesTeaMaybeWithMilkAction
  extend LightService::Action
  expects :tea, :use_milk, :maybe => [:milk]
  promises :milk_tea
end
```

* A fourth option would be to remove `maybe` altogether and force all actions to access their `expected` keys.  That would encourage fewer conditionals in actions, but might be too onerous a restriction to place on users.  
I don't think this should work with `promises`, because I like the idea that executing an action will always add specific keys to the context, and LS will raise an error if that contract is violated.  Allowing actions to only sometimes add keys to the context would complicate any actions that are executed later, because now those actions have to code defensively in case the upstream action didn't add a key to the context.